src/qs8-vadd/gen/qs8-vadd-minmax-hvx-u128.c

```diff
@@ -0,0 +1,137 @@
+  assert(input_b != NULL);
+  assert(output != NULL);
+
+  const HVX_Vector vbias = Q6_V_vsplat_R(*((int32_t *) &params->hvx.bias));
+  const HVX_Vector va_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.a_multiplier));
+  const HVX_Vector vb_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.b_multiplier));
+  const int32_t first_shift = params->hvx.first_shift;
+  const int32_t rest_shift = params->hvx.rest_shift;
+  const HVX_Vector voutput_zero_point = Q6_Vh_vsplat_R(*((int16_t *) &params->hvx.output_zero_point));
+  const HVX_Vector voutput_min = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_min));
+  const HVX_Vector voutput_max = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_max));
+  int8_t* ptr_o = output;
+
+  for (; batch >= 128 * sizeof(int8_t); batch -= 128 * sizeof(int8_t)) {
+    HVX_Vector va0 = *((HVX_UVector*)input_a);
+    HVX_Vector vb0 = *((HVX_UVector*)input_b);
+    input_a += 128;
+    input_b += 128;
+
+    // widen: 8-bit to 16-bit
+    HVX_VectorPair va0_i16 = Q6_Wh_vsxt_Vb(va0);
+    HVX_Vector va0_i16_even = Q6_V_lo_W(va0_i16);
+    HVX_Vector va0_i16_odd = Q6_V_hi_W(va0_i16);
+
+    HVX_VectorPair vb0_i16 = Q6_Wh_vsxt_Vb(vb0);
+    HVX_Vector vb0_i16_even = Q6_V_lo_W(vb0_i16);
+    HVX_Vector vb0_i16_odd = Q6_V_hi_W(vb0_i16);
+
+    // vacc = va * va_multiplier + vb * vb_multiplier with widening to 32-bit
+    HVX_VectorPair va0_mul_even = Q6_Vw_vmpyi_VwVh(va_multiplier, va0_i16_even);
+    HVX_VectorPair va0_mul_odd = Q6_Vw_vmpyi_VwVh(va_multiplier, va0_i16_odd);
+    HVX_VectorPair vb0_mul_even = Q6_Vw_vmpyi_VwVh(vb_multiplier, vb0_i16_even);
+    HVX_VectorPair vb0_mul_odd = Q6_Vw_vmpyi_VwVh(vb_multiplier, vb0_i16_odd);
+
+    HVX_Vector vacc0_even = Q6_Vw_vadd_VwVw(Q6_V_lo_W(va0_mul_even), Q6_V_lo_W(vb0_mul_even));
+    HVX_Vector vacc0_odd = Q6_Vw_vadd_VwVw(Q6_V_lo_W(va0_mul_odd), Q6_V_lo_W(vb0_mul_odd));
+    HVX_Vector vacc1_even = Q6_Vw_vadd_VwVw(Q6_V_hi_W(va0_mul_even), Q6_V_hi_W(vb0_mul_even));
+    HVX_Vector vacc1_odd = Q6_Vw_vadd_VwVw(Q6_V_hi_W(va0_mul_odd), Q6_V_hi_W(vb0_mul_odd));
+
+    // vacc = vbias + vacc
+    vacc0_even = Q6_Vw_vadd_VwVw(vbias, vacc0_even);
+    vacc0_odd = Q6_Vw_vadd_VwVw(vbias, vacc0_odd);
+    vacc1_even = Q6_Vw_vadd_VwVw(vbias, vacc1_even);
+    vacc1_odd = Q6_Vw_vadd_VwVw(vbias, vacc1_odd);
+
+    // vacc = shift and narrow to 16-bit. Add voutput_zero_point
+    HVX_Vector vacc0 = Q6_Vh_vasr_VwVwR_sat(vacc0_odd, vacc0_even, first_shift);
+    vacc0 = Q6_Vh_vadd_VhVh(voutput_zero_point, Q6_Vh_vasr_VhR(vacc0, rest_shift));
+    HVX_Vector vacc1 = Q6_Vh_vasr_VwVwR_sat(vacc1_odd, vacc1_even, first_shift);
+    vacc1 = Q6_Vh_vadd_VhVh(voutput_zero_point, Q6_Vh_vasr_VhR(vacc1, rest_shift));
+
+    // narrow: 16-bit to 8-bit
+    HVX_Vector vout0 = Q6_Vb_vpack_VhVh_sat(vacc1, vacc0);
+
+    // minmax
+    vout0 = Q6_Vb_vmax_VbVb(voutput_min, vout0);
+    vout0 = Q6_Vb_vmin_VbVb(voutput_max, vout0);
+
+    // store output
+    Q6_V_vstu_variable(ptr_o, 128, vout0);
```

Comment:
Shouldn't this just be a regular vector store?

Reply:
Changed!